tests/functional/base.py

```diff
@@ -0,0 +1,316 @@
+            return -1
+        finally:
+            self.client_info['admin'] = admin_auth
+        return 0
+
+    def _init_client(self, client):
+        auth = self.client_info[client]
+        self.clients[client] = EsiCliClient(
+                cli_dir=self.config['cli_dir'],
+                username=auth['username'],
+                password=auth['password'],
+                tenant_name=auth['project_name'],
+                user_domain_name=auth['user_domain_name'],
+                project_domain_name=auth['project_domain_name'],
+                identity_api_version='3',
+                uri=auth['auth_url'])
+        
+    def _init_dummy_cloud(self, name, role):
+        if role not in ('owner', 'lessee'):
+            raise Exception('unknown role: %s' % role)
+
+        cloud = {}
+        for field in 'project', 'user':
+            cloud[field] = { 'name': data_utils.rand_name('esi-%s-%s'
+                    % (field, name)) }
+
+        def initialize(field, params):
+            output = self.clients['admin'].openstack('%s create %s' %
+                    (field, cloud[field]['name']), '', params)
+            cloud[field]['id'] = self.parse_details(self, output)['id']
+
+        pw = data_utils.rand_password()
+        initialize('project', '--domain default --enable')
+        initialize('user', '--domain default --password %s --enable' % pw)
+        self.clients['admin'].openstack('role add', '',
+                '--user %s --project %s esi_leap_%s' % (cloud['user']['name'],
+                    cloud['project']['name'], role))
+
+        self.client_info[name] = {
+                'auth_type': 'password',
+                'auth_url': self.client_info['admin']['auth_url'],
+                'username': cloud['user']['name'],
+                'user_id': cloud['user']['id'],
+                'password': pw,
+                'project_name': cloud['project']['name'],
+                'project_id': cloud['project']['id'],
+                'user_domain_name': 'default',
+                'project_domain_name': 'default'}
+
+        for field in cloud.keys():
+            self._cleanups.append([self.clients['admin'].openstack, 
+                    '%s delete' % field, '', '%s' % cloud[field]['name']])
+        self._cleanups.append([self.clients['admin'].openstack, 'role remove', '',
+                '--user %s --project %s esi_leap_%s' % (cloud['user']['name'],
+                    cloud['project']['name'], role)])
+        return 0
+
+    def _init_roles(self):
+        for role in 'esi_leap_owner', 'esi_leap_lessee':
+            try:
+                self.clients['admin'].openstack('role show', '', '%s' % role)
+            except:
+                self.clients['admin'].openstack('role create', '', '%s' % role)
+                self._cleanups.append([self.clients['admin'].openstack,
+                    'role delete', '', '%s' % role])
+
+    def _new_dummy_node(self, owner):
```

Comment:
This feels like it should be split off into something like `utils/dummy_node.py`

Reply:
Good call. It used to be called by `setUp()` in an early implementation but since it isn't now I'll move it out.